src/registered.rs

```diff
@@ -0,0 +1,345 @@
+//! The instantiation of the [hardened key derivation framework] in this module may be
+//! used for this purpose.
+//!
+//! It is strongly RECOMMENDED that implementors ensure that documentation of the
+//! usage and derivation paths of the application protocol's key tree in the
+//! corresponding ZIP is substantially complete, before public deployment of software
+//! or hardware using this mechanism. The ZIP process allows for subsequent updates
+//! and corrections.
+//!
+//! The functionality of this module is similar to that of the [`zip32::arbitrary`]
+//! module, with the following improvements:
+//!
+//! - The key tree is associated with the ZIP that should document it, and cannot
+//!   collide with the tree for any other ZIP.
+//! - Child indices may include byte sequence tags.
+//! - A 64-bit cryptovalue can be derived at the same path as any node in the tree,
+//!   without any cryptographic unsafety.
+//!
+//! The keys derived by the functions in this module will be unrelated to any keys
+//! derived by functions in the [`zip32::arbitrary`] module, even if the same context
+//! string and seed are used.
+//!
+//! Defined in [ZIP 32: Registered key derivation][regkd].
+//!
+//! [hardened key derivation framework]: crate::hardened_only
+//! [regkd]: https://zips.z.cash/zip-0032#specification-registered-key-derivation
+//! [ZIP process]: https://zips.z.cash/zip-0000
+//! [`zip32::arbitrary`]: `crate::arbitrary`
+
+use zcash_spec::PrfExpand;
+
+use crate::{
+    hardened_only::{Context, HardenedOnlyCkdDomain, HardenedOnlyKey},
+    ChainCode, ChildIndex,
+};
+
+use super::with_ikm;
+
+struct Registered;
+
+impl Context for Registered {
+    const MKG_DOMAIN: [u8; 16] = *b"ZIPRegistered_KD";
+    const CKD_DOMAIN: HardenedOnlyCkdDomain = PrfExpand::REGISTERED_ZIP32_CHILD;
+}
+
+/// A registered extended secret key.
+///
+/// Defined in [ZIP 32: Registered key derivation][regkd].
+///
+/// [regkd]: https://zips.z.cash/zip-0032#specification-registered-key-derivation
+pub struct SecretKey {
+    inner: HardenedOnlyKey<Registered>,
+}
+
+impl SecretKey {
+    /// Derives a key for a registered application protocol at the given path from the
+    /// given seed. Each path element may consist of an index and (possibly empty) tag.
+    ///
+    /// - `zip_number` is the number of the ZIP defining the application protocol.
+    ///   The corresponding hardened index (with empty tag) will be prepended to the
+    ///   `subpath` to obtain the full derivation path.
+    /// - `context_string` is an identifier for the context in which this key will be
+    ///   used. It must be globally unique.
+    ///
+    /// # Panics
+    ///
+    /// Panics if:
+    /// - the context string is empty or longer than 252 bytes.
+    /// - the seed is shorter than 32 bytes or longer than 252 bytes.
+    pub fn from_subpath(
+        context_string: &[u8],
+        seed: &[u8],
+        zip_number: u16,
+        subpath: &[(ChildIndex, &[u8])],
```

Comment:
The `subpath` parameter is insufficiently documented. Also, I would prefer the elements of the path be a bespoke struct instead of a bare tuple.